src/nupic/bindings/algorithms.i

```diff
@@ -1163,10 +1163,22 @@ void forceRetentionOfImageSensorLiteLibrary(void) {
       self._initFromCapnpPyBytes(proto.as_builder().to_bytes()) # copy * 2
   %}
 
-  inline void compute(PyObject *py_x, bool learn, PyObject *py_y)
+  inline void compute(PyObject *py_inputArray, bool learn, PyObject *py_activeArray)
   {
-    PyArrayObject* x = (PyArrayObject*) py_x;
-    PyArrayObject* y = (PyArrayObject*) py_y;
+    PyArrayObject* x = (PyArrayObject*) py_inputArray;
+    PyArrayObject* y = (PyArrayObject*) py_activeArray;
+    if (!PyArray_ISUNSIGNED(x) || PyArray_DTYPE(x)->elsize != sizeof(nupic::UInt))
+    {
+      NTA_THROW << "Invalid data type given for input array."
+                << " Expecting 'uint" << sizeof(nupic::UInt) * 8 << "'"
+                << " but got '" << PyArray_DTYPE(x)->type << "'";
+    }
+    if (!PyArray_ISUNSIGNED(y) || PyArray_DTYPE(y)->elsize != sizeof(nupic::UInt))
+    {
+      NTA_THROW << "Invalid data type given for active array."
+                << " Expecting 'uint" << sizeof(nupic::UInt) * 8 << "'"
+                << " but got '" << PyArray_DTYPE(y)->type << "'";
+    }
```

Comment:
It might be worth creating a `CheckedNumpyVectorWeakRefT` by copy-pasting the `NumpyVectorWeakRefT` in https://github.com/numenta/nupic.core/blob/91b45451ec77aabbf71ee9f3645ec704532664f1/src/nupic/py_support/NumpyVector.hpp#L411 and changing it to do checks in release builds. Then we can reuse it for any dtype. The existing code would check that the dtype is correct (which, strictly speaking, this code doesn't currently do). I like that you throw useful error text -- we'd want that in the Checked class.

This we could change this code to

~~~c++
nupic::CheckedNumpyVectorWeakRefT<nupic::UInt32> inputArray(py_inputArray);
nupic::CheckedNumpyVectorWeakRefT<nupic::UInt32> activeArray(py_activeArray);
self->compute(inputArray.begin(), learn, activeArray.begin());
~~~

Reply:
@mrcslws regarding 
> The existing code would check that the dtype is correct (which, strictly speaking, this code doesn't currently do)

Can you give an example where this code does not check the  `dtype` of the numpy array passed to the function? I would like to cover that condition with a test.

Reply:
@mrcslws Why create the new `CheckedNumpyVectorWeakRefT` class?
`NumpyVectorWeakRefT` seems to do what I need. 
Any reason why this class should only throw an exception in debug mode?

Reply:
I intentionally used `NTA_ASSERT` in the `NumpyVectorWeakRefT`. In the SparseMatrix, the parameters are checked in the Python bindings. It would be wasteful to recheck them. The main purpose of the class isn't to check its parameters, it's to expose the array data.